### ISSUE TITLE
Billboard sizeInMeters clamped to 65 kibi-meters

### ISSUE THREAD

**Issue by thw0rted** [CesiumGS/cesium#9047]
[Sandcastle example](https://sandcastle.cesium.com/#c=vVVdT9swFP0rVrSHlAU3DYHyUdC2UqFJsEkUjZdIw00uqYVjV7aTFlD/+5w4TUNH2TSh9aGKj889995zHacgEhUU5iDRKZpTnog5btYc5mgIiuYZ/lFhbuTE1XoouCaUg4wcDz2rGDhciQSOV/TxCsHj4ejbKDhfdk4iHvFYcKVRTHhB1Dphs05EnGfANY4lEA0jBuWqzFoxIqdUsc94ThM9NTGh76/BKdB0qg0aWNSm0wuD1IwUdFk8LErZILGShqkXWGkpHmCsHxkYfuTMcjljEDkndpuZfm/rpD2/BjNRwI1wfc/vtGgGaVfpobd3XxS/wfT/uN1o2/rdttMSYt0z9dqRYmMm1RQUJkniPkccoZlQBhG8mdyQSG2eCN/D91Jk55BKAOXu9vuesbrjlUGlKuEpM/OuRBCKhZAJ5WZmqlG6XrE2hXDg7R3hI2+3f4APvdDHPSuLUGYEJCVsXY1gQuKb0efLirGs/ieUsYkgMmnSK/oEX/kVmGiTX8scasHK4GMUhuG+X0PWxmMUmN9+jdGMpKYZa7TNFPHlppHBOxqJ99/HyrD0b6+y8qB87P8XK9EOCl6zs4W/ZWnEaxufhMjMId4wte07KdJrktC8vB3cupMRY3SmBE3w7cX4MMQZ5TTLs5r3EW2jkcWa1kFdFNQ3hDC+MpG6d43NqIeoQh+eq9cHNzPCrdngnwVhOdTX0M660CVa/F1kfVe1Q+86myV9WQ3oZUnN3Gz+ds71lk1gRbd2Gqxlg3/uNHjXTl+UtLXTYGunjucMVHmLn9mjiNAnms2E1CiXzMW4qyGbsbLA7iSPH8B8blR16ErqoNsOHSS0QDQ5feW7h2JGlDI79zljY/PWRM7ZoGv4v4UyYXrm6fcCJCOPJW3aO7u0IMZ40DXL1yO1EGxC5IbyLw) -- the purple border should line up with the edges of the teal fill, and it does for the smaller shape (44km) but not the larger one (88km).

Per [discussion on the forum](https://community.cesium.com/t/what-on-earth-is-going-on-with-billboardgraphics-sizeinmeters/10260/2?u=james_b) it looks like at some point the billboard width is being clamped to 65535.    I'm guessing height has the same problem.  The clamping should be applied to the actual pixel count of the texture, not the logical size.

**Comment by hpinkos** [comment]
Yes, as pointed out in the forum discussion, this is due to the way we're compressing billboard attributes to send to the vertex shader for rendering. Height will have the same limitations.

https://github.com/CesiumGS/cesium/blob/master/Source/Scene/BillboardCollection.js#L1246

Because a scene can potentially have a very large number of billboards, we have to be very careful about the amount of memory used in order to optimize performance.  There's also a limit to the number of varying attributes we can send to the shader so we have to pack multiple values into one attribute.  I don't know that there's a very easy fix to this.  We might want to throw a `DeveloperError` in the meantime if the values are larger than expected since this seems like a not-very-common use case.

**Comment by thw0rted** [comment]
I don't think I understand.  The memory usage should be based on input texture size in pixels, and shouldn't depend on the ("output") rendered texture size in world coordinates, right?  Or, you're saying that to conserve memory, the world-coordinate size is being passed as 16-bit packed values?  And no other existing use cases exceed 16-bit world-coordinate sizes so it's not a priority?

As for how to handle it, I assume that other shapes are broken up into smaller primitives -- could the billboard texture be applied to a larger number of sub-64k rectangles?

**Comment by hpinkos** [comment]
It has to do with GPU memory, not texture size.  Every billboard allocates these attributes in the vertex shader: https://github.com/CesiumGS/cesium/blob/master/Source/Shaders/BillboardCollectionVS.glsl#L4
To conserve memory, the width/height are packed as a 16-bit value.  Compressing these and other values lets us use one vec4 attribute to represent 6 or 7 different values instead of having 6 or 7 separate attributes.

That being said, I did take a closer look and we do have some free bits in one of the compressed attributes we could use to allow for larger width/height values.  Or we could decide add another attribute so we don't compress width/height.  We'll just have to keep performance in mind since in the most common use case (when `sizeInMeters = false) width/height are pixel size and will be much smaller than 65535.

**Comment by javagl** [comment]
So if I understood this correctly, then the point of this compression is that instead of storing the sizes of 1024 billboards as 4 byte + 4 byte (taking 8KB), the sizes are stored as 2 byte + 2 byte (taking 4KB, plus advantages for passing the data to the shader).

> As for how to handle it, I assume that other shapes are broken up into smaller primitives -- could the billboard texture be applied to a larger number of sub-64k rectangles?

I'm not familiar with the specific API, but from quickly skimming over it: A workaround *might* be to split a 100000x50000 billboard into two parts (each being smaller than 65k then), and using [`Billboard.html#setImageSubRegion`](https://cesium.com/docs/cesiumjs-ref-doc/Billboard.html#setImageSubRegion) to let them show the relevant part of the image. For *really* large Billboards (where you'd have to split it into, say 6x4 sub-Billboards), this might be a bit clumsy, but could probably be easily wrapped into a function that takes an image and a size and generates the required sub-Billboards, maybe directly as a Billboard collection.
